python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+            The covariance matrix
+        """
+        covariance = np.dot(self.W_.T, self.W_) + np.diag(self.sigmas_)
+        return covariance
+
+    def _create_prior(self):
+        """
+        This creates the function representing prior on pararmeters
+
+        Parameters
+        ----------
+        log_prior : function
+            The function representing the negative log density of the prior
+        """
+        pd = self.prior_options
+
+        def log_prior(trainable_variables):
+            list_gamma_log_probs = []
+            for k in self.n_groups:
+                list_gamma_log_probs.append(
+                    Gamma(pd["alpha"], pd["beta"]).log_prob(
+                        trainable_variables[k + 1]
+                    )
+                )
+
+            gamma_log_probs = torch.stack(list_gamma_log_probs)
+            return torch.mean(gamma_log_probs)
+
+        return log_prior
+
+    def _fill_prior_options(self, prior_options):
+        """
+        Fills in options for prior parameters
+
+        Paramters
+        ---------
+        new_dict : dictionary
+            The prior parameters used to specify the prior
+        """
+        default_dict = {"alpha": 3.0, "beta": 3.0}
+        new_dict = {**default_dict, **prior_options}
+        return new_dict
+
+    def _initialize_variables(self, X):
+        """
+        Initializes the variables of the model
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : tf.Variable-like,(n_components,p)
+            The loadings of our latent factor model
+
+        sigmal_ : list
+            A list of the isotropic noises for each group
+        """
+        model = dp.PCA(self.n_components)
+        S_hat = model.fit_transform(X)
+        W_init = model.components_.astype(np.float32)
+        W_ = torch.tensor(W_init, requires_grad=True)
+        X_recon = np.dot(S_hat, W_init)
+        diff = np.mean((X - X_recon) ** 2)
+        sinv = softplus_inverse_np(diff * np.ones(1).astype(np.float32))
+        sigmal_ = [
+            torch.tensor(sinv, requires_grad=True) for i in range(self.n_groups)
+        ]
+        return W_, sigmal_
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of variables learned by pytorch
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigmas_ : np.array-like,(p,)
+            The diagonal variances
+        """
+        self.W_ = trainable_variables[0].detach().numpy()
+        self.sigmas_ = np.zeros(self.p)
+        for k in range(self.n_groups):
+            sigma2 = nn.Softplus()(trainable_variables[k + 1])
+            self.sigmas_[self.groups == k] = sigma2
+
+    def _test_inputs(self, X, groups):
```

Comment:
same comment as before about the name of this method.  In fact, since it doesn't really depend on the instance, it can just be a pure helper function instead of being reimplemented in every class

Reply:
No these are different even here, but especially when we start adding supervision or adversarial methods to account for undesired variables.